Add classname to "Required class information is missing" error

See https://github.com/wildfly/jandex/blob/7f2c075ba9246e4b6cc8fe7f3e3678dc2b64891c/src/main/java/org/jboss/jandex/Indexer.java#L926

The error message is hard to understand, and the root cause is not clear.

What do you think if the error includes the class name or something else for making the exception actionable?

The current exception:

```
Caused by: java.lang.IllegalStateException: Required class information is missing
         at org.jboss.jandex.Indexer.rebuildNestedType(Indexer.java:926)
         at org.jboss.jandex.Indexer.resolveTypePath(Indexer.java:786)
         at org.jboss.jandex.Indexer.resolveTypeAnnotation(Indexer.java:705)
         at org.jboss.jandex.Indexer.resolveTypeAnnotations(Indexer.java:613)
         at org.jboss.jandex.Indexer.index(Indexer.java:1602)
         at org.hibernate.boot.archive.scan.spi.ClassFileArchiveEntryHandler.toClassDescriptor(ClassFileArchiveEntryHandler.java:64)
         at org.hibernate.boot.archive.scan.spi.ClassFileArchiveEntryHandler.handleEntry(ClassFileArchiveEntryHandler.java:52)
```


Context: https://www.postgresql.org/message-id/AM0PR04MB652919E6F344C84DC3073D0D96140%40AM0PR04MB6529.eurprd04.prod.outlook.com

Another exception:

```
            Caused by: java.lang.IllegalStateException: Class extends type annotation appeared on a non class target
                at org.jboss.jandex.Indexer.processTypeAnnotation(Indexer.java:538)
                at org.jboss.jandex.Indexer.processTypeAnnotations(Indexer.java:515)
                at org.jboss.jandex.Indexer.processAttributes(Indexer.java:355)
                at org.jboss.jandex.Indexer.processMethodInfo(Indexer.java:306)
                at org.jboss.jandex.Indexer.index(Indexer.java:1602)
```

--- a/pom.xml
+++ b/pom.xml
@@ -131,7 +131,7 @@
         <dependency>
             <groupId>org.jboss.jandex</groupId>
             <artifactId>typeannotation-test</artifactId>
-            <version>1.5</version>
+            <version>1.6</version>
             <type>jar</type>
             <scope>test</scope>
         </dependency>
@@ -158,7 +158,7 @@
                 <enabled>true</enabled>
             </releases>
             <snapshots>
-                <enabled>false</enabled>
+                <enabled>true</enabled>
             </snapshots>
         </repository>
     </repositories>

--- a/src/main/java/org/jboss/jandex/Indexer.java
+++ b/src/main/java/org/jboss/jandex/Indexer.java
@@ -985,8 +985,16 @@ public final class Indexer {
             }
         }
 
+        // UGLY HACK - Java 8 has a compiler bug that inserts bogus extra NESTED values
+        // when the enclosing type is an anonymous class (potentially multiple if there
+        // is multiple nested anonymous classes). Java 11 does not suffer from this issue
+        if (depth > 0 && hasAnonymousEncloser(typeAnnotationState)) {
+            return resolveTypePath(type, typeAnnotationState);
+        }
+
         if (last == null) {
-            throw new IllegalStateException("Required class information is missing");
+            throw new IllegalStateException("Required class information is missing on: "
+                    + typeAnnotationState.target.enclosingTarget().asClass().name().toString());
         }
 
         return last;
@@ -1015,7 +1023,18 @@ public final class Indexer {
             }
         }
 
+        // UGLY HACK - see comment in rebuildNestedType
+        if (hasAnonymousEncloser(typeAnnotationState)) {
+            return searchTypePath(type, typeAnnotationState);
+        }
+
         throw new IllegalStateException("Required class information is missing");
+    }
+
+    private boolean hasAnonymousEncloser(TypeAnnotationState typeAnnotationState) {
+        return typeAnnotationState.target instanceof ClassExtendsTypeTarget
+                && typeAnnotationState.target
+                .enclosingTarget().asClass().nestingType() == ClassInfo.NestingType.ANONYMOUS;
     }
 
     private ArrayDeque<InnerClassInfo> buildClassesQueue(DotName name) {

--- a/src/test/java/org/jboss/jandex/test/TypeParameterBoundTestCase.java
+++ b/src/test/java/org/jboss/jandex/test/TypeParameterBoundTestCase.java
@@ -72,6 +72,34 @@ public class TypeParameterBoundTestCase {
                 info.typeParameters().get(0).toString());
     }
 
+
+    @Test
+    public void classExtendsOnInner() throws IOException {
+        Indexer indexer = new Indexer();
+        ClassInfo info = indexer.index(getClassBytes("test/TypeParameterBoundExample$IteratorSupplier.class"));
+        Assert.assertEquals("java.util.function.Supplier<java.util.function.Consumer<@Nullable java.lang.Object[]>>",
+                            info.interfaceTypes().get(0).toString());
+    }
+
+    @Test
+    public void classExtendsOnAnonInInner() throws IOException {
+        Indexer indexer = new Indexer();
+        ClassInfo info = indexer.index(getClassBytes("test/TypeParameterBoundExample$IteratorSupplier$1.class"));
+        Assert.assertEquals("java.util.function.Consumer<@Nullable java.lang.Object[]>",
+                           info.interfaceTypes().get(0).toString());
+    }
+
+    @Test
+    public void classExtendsNestAnonExtendsInner() throws IOException {
+        Indexer indexer = new Indexer();
+        ClassInfo info = indexer.index(getClassBytes("test/TypeParameterBoundExample$Nest1$Nest2$Nest3$1$1.class"));
+        Assert.assertEquals("test.TypeParameterBoundExample$Nest1<java.lang.String>.Nest2<java.lang.Object[]>.Nest3<@Nullable java.lang.Integer>",
+                    info.superClassType().toString());
+        info = indexer.index(getClassBytes("test/TypeParameterBoundExample$Nest1$Nest2$Nest3$1$2.class"));
+        Assert.assertEquals("test.TypeParameterBoundExample$Nest1<java.lang.String>.Nest2<@Nullable java.lang.Object[]>.Nest3<java.lang.Integer>",
+                           info.superClassType().toString());
+    }
+
     @Test
     public void serializableListConsumerDA() throws IOException {
         Indexer indexer = new Indexer();

